File: 2024/1614B.go
// created: 2024/6/8 23:45:18 author:  https://github.com/funcdfs
// problem: https://codeforces.com/contest/1614/problem/B

// #region import
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
)

// #endregion import

// solve -------------------------------------------------------------

func solve(_case int) {
	// fmt.Fprintln(os.Stderr, "# CASE: ", _case)
	n := input[int]()
	a := make([][2]int, n)
	for i := range a {
		a[i][0] = input[int]()
		a[i][1] = i + 1
	}
	sort.Slice(a, func(i, j int) bool {
		return a[i][0] > a[j][0]
	})
	x := make([]int, n+1)
	x[0] = 0
	ans, l, r := int64(0), -1, 1
	for i := 0; i < n; i++ {
		id := a[i][1]
		if r == -l {
			x[id] = l
			l -= 1
		} else {
			x[id] = r
			r += 1
		}
		ans += int64(abs(x[id])) * int64(a[i][0])
	}
	print(ans * 2)
	print(x...)
}
func abs(x int) int {
	if x < 0 {
		x = -x
	}
	return x
}
func preProcess() {

}

// solve -------------------------------------------------------------

// #region main
func main() {
	_in = bufio.NewReader(os.Stdin)
	_out = bufio.NewWriter(os.Stdout)
	defer _out.Flush()
	preProcess()
	testCaseCnt := input[int]()
	for i := 0; i < testCaseCnt; i++ {
		solve(i + 1)
	}
}

// #endregion main

// #region fastIO
var _in *bufio.Reader
var _out *bufio.Writer

func input[T any]() T {
	var value T
	fmt.Fscan(_in, &value)
	return value
}
func inputSlice[T any](length int) []T {
	data := make([]T, length)
	for i := 0; i < length; i++ {
		data[i] = input[T]()
	}
	return data
}
func print[T any](x ...T) {
	for i := range x {
		fmt.Fprint(_out, x[i])
		if i == len(x)-1 {
			fmt.Fprintln(_out)
		} else {
			fmt.Fprint(_out, " ")
		}
	}
}

// #endregion fastIO

File: 2024/1614B.cpp
/**
 * name: B. Divan and a New Project
 * link: https://codeforces.com/contest/1614/problem/B
 * time: 2024/6/8 23:23:16 https://github.com/funcdfs
 **/

#pragma region github_funcdfs
#include <bits/stdc++.h>
using namespace std;
#define endl '\n'
#ifdef LOCAL
#include "algo/dbg.h"
#else
#define dbg(...) 42
#endif
#pragma endregion github_funcdfs

auto solve(int _case) -> void {
   dbg(_case);

   int n = 0;
   cin >> n;
   vector<pair<int, int>> a(n);
   for (int i = 0; i < n; i++) {
      cin >> a[i].first;
      a[i].second = i + 1;
   }
   sort(a.begin(), a.end(), [&](auto& x, auto& y) {
      return x.first > y.first;
   });
   int l = -1, r = 1;
   vector<int> x(n + 1, 0);
   x[0] = 0;
   int64_t ans = 0;
   for (int i = 0; i < n; i++) {
      int id = a[i].second;
      if (r == -l) {
         x[id] = l;
         l -= 1;
      } else {
         x[id] = r;
         r += 1;
      }
      ans += (int64_t)abs(x[id]) * (int64_t)a[i].first;
   }
   cout << ans * 2 << endl;
   for (int i = 0; i <= n; i++) {
      cout << x[i] << " \n"[i == n];
   }
   return;
}

/*
第 i 个建筑访问 a[i] 次，每次访问需要花费 2 x[x0 - xi] 的时间来走路。xi 表示第 i
个建筑的位置。 求出最少需要走多远，以及一个具体的 x 数组。
*/

#pragma region main func
auto main() -> int32_t {
   cin.tie(nullptr), cout.tie(nullptr), ios::sync_with_stdio(false);
   int testCaseCnt = 0;
   cin >> testCaseCnt;
   for (int _case = 1; _case <= testCaseCnt; _case++) {
      solve(_case);
   }
   return 0;
}
#pragma endregion main func
